File: 0145_binaryTreePostorderTraversal/main.py
from typing import Optional, List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        def postorder(node):
            if not node:
                return
            postorder(node.left)
            postorder(node.right)
            res.append(node.val)

        postorder(root)
        
        return res
    
class Solution1:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        stack = []
        last = None
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                node = stack[-1]
                if node.right and node.right != last:
                    root = root.right
                else:
                    res.append(node.val)
                    last = node
                    stack.pop()
        return res
    
class Solution2:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        
        stack1 = []
        stack2 = []
        res = []
        stack1.append(root)
        while stack1:
            root = stack1.pop()
            if root.left:
                stack1.append(root.left)
            if root.right:
                stack1.append(root.right)

            stack2.append(root)
        
        while stack2:
            res.append(stack2[-1].val)
            stack2.pop()

        return res
    

File: 0145_binaryTreePostorderTraversal/main.cpp
#include <vector>
#include <stack>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    std::vector<int> postorderTraversal(TreeNode* root) {
        std::vector<int> res;
        inorder(root, res);
        return res;
    }
private:
    void inorder(TreeNode* root, std::vector<int>& res) {
        if (!root) {
            return;
        }
        inorder(root -> left, res);
        inorder(root -> right, res);
        res.push_back(root -> val);
    }
};

class Solution1 {
public:
    std::vector<int> postorderTraversal(TreeNode* root) {
        std::vector<int> res;
        std::stack<TreeNode*> st;
        TreeNode* last = NULL;
        while (root || !st.empty()) {
            if (root) {
                st.push(root);
                root = root -> left;
            } else {
                TreeNode* node = st.top();
                if (node -> right && last != node -> right) {
                    root = node -> right;
                } else {
                    res.push_back(node -> val);
                    last = node;
                    st.pop();
                }
            }
        }
        return res;
    }
};


class Solution2 {
public:
    std::vector<int> postorderTraversal(TreeNode* root) {
        if(!root)
            return {};

        std::stack<TreeNode*> s1;
        std::stack<TreeNode*> s2;
        std::vector<int> res;
        s1.push(root);
        while(!s1.empty()){
            root=s1.top();
            s1.pop();

            if(root->left)
                s1.push(root->left);
            
            if(root->right!=NULL)
                s1.push(root->right);
            
            s2.push(root);
        }
        while(!s2.empty()){
            res.push_back(s2.top()->val);
            s2.pop();
        }
        return res;
    }
};